design-decisions/2018-09-18-Substitution-Predicate-Top-Evaluation.md

```diff
@@ -0,0 +1,242 @@
+at the current evaluation level (`φ`) or any level above it in the context `C`.
+This means that having a bottom `term` does not force the entire pair to bottom,
+but a bottom predicate does, since its branch will be removed when doing the
+actual split.
+
+Note that at the splitting level, the `(term, predicate ∧ substitution)` pair
+has the original meaning of `(term ∧ predicate ∧ substitution)`.
+
+Function evaluation
+-------------------
+
+Say we have a function defined as
+```
+f(x)=v1 if P1
+f(x)=v2 if P2
+...
+f(x)=vn if Pn
+```
+
+Then, when evaluating f inside a context `C[f(y)]` we do this:
+```
+C[f(y)]
+    = (P1 ∧ C[f(y)]) ∨ (¬P1 ∧ C[f(y)])
+    = (P1 ∧ C[f(y)])
+        ∨ (¬P1 ∧ P2 ∧ C[f(y)])
+        ∨ (¬P1 ∧ ¬P2 ∧ C[f(y)])
+    ...
+    = (P1 ∧ C[f(y)])
+        ∨ (¬P1 ∧ P2 ∧ C[f(y)])
+        ∨ (¬P1 ∧ ¬P2 ∧ P3 ∧ C[f(y)])
+        ...
+        ∨ (¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ Pn ∧ C[f(y)])
+        ∨ (¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ ¬Pn ∧ C[f(y)])
+```
+which, assuming confluence or disjoint predicates, and using
+`P ∧ C[φ] = P ∧ C[φ ∧ P])` (see the proof-ish section) and `f(y)=⊥` if no
+function definition branch is matched (i.e. `¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ ¬Pn`,
+probabily needs a separate axiom), becomes
+```
+C[f(y)]
+    = (P1 ∧ C[f(y) ∧ P1])
+        ∨ (P2 ∧ C[f(y) ∧ P2])
+        ∨ (P3 ∧ C[f(y) ∧ P3])
+        ...
+        ∨ (Pn ∧ C[f(y) ∧ Pn])
+        ∨ (¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ ¬Pn ∧ C[⊥])
+    = (P1 ∧ C[v1])
+        ∨ (P2 ∧ C[v2])
+        ∨ (P3 ∧ C[v3])
+        ...
+        ∨ (Pn ∧ C[vn])
+        ∨ (¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ ¬Pn ∧ C[⊥])
+```
+
+But, if we want to do bottom-up splitting, we can also evaluate it like this:
+```
+f(y)
+    = (f(y) ∧ P1) ∨ (f(y) ∧ ¬P1)
+    = (f(y) ∧ P1)
+        ∨ (f(y) ∧ ¬P1 ∧ P2)
+        ∨ (f(y) ∧ ¬P1 ∧ ¬P2)
+    ...
+    = (f(y) ∧ P1)
+        ∨ (f(y) ∧ ¬P1 ∧ P2)
+        ∨ (f(y) ∧ ¬P1 ∧ ¬P2 ∧ P3)
+        ...
+        ∨ (f(y) ∧ ¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ Pn)
+        ∨ (f(y) ∧ ¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ ¬Pn)
+```
+By using confluence or disjoint predicates, this becomes
+```
+f(y)
+    = (f(y) ∧ P1)
+        ∨ (f(y) ∧ P2)
+        ∨ (f(y) ∧ P3)
+        ...
+        ∨ (f(y) ∧ Pn)
+        ∨ (f(y) ∧ ¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ ¬Pn)
+    = (v1 ∧ P1)
+        ∨ (v2 ∧ P2)
+        ∨ (v3 ∧ P3)
+        ...
+        ∨ (vn ∧ Pn)
+        ∨ (⊥ ∧ ¬P1 ∧ ¬P2 ... ∧ ¬Pn-1 ∧ ¬Pn)
+    = (v1 ∧ P1)
+        ∨ (v2 ∧ P2)
+        ∨ (v3 ∧ P3)
+        ...
+        ∨ (vn ∧ Pn)
+```
+When evaluating this inside the context, we have (assuming confluence, see
+the proof-ish section for details)
+```
+C[f(y)]
+    = C[(v1 ∧ P1) ∨ (v2 ∧ P2) ∨ (v3 ∧ P3) ... ∨ (vn ∧ Pn)]
+    = (P1 ∧ C[v1])
+        ∨ (P2 ∧ C[v2])
+        ∨ (P3 ∧ C[v3])
+        ...
+        ∨ (Pn ∧ C[vn])])
+        ∨ (¬P1 ∧ ¬P2 ∧ ... ∧ ¬Pn ∧ C[⊥])
+```
+
+This means that we are justified in representing the function evaluation
+using the pair representation:
+```
+f(x) = (v1, P1 ∧ []) ∨ (v2, P2 ∧ []) ∨ ... ∨ (vn, Pn ∧ [])
+        ∨ (⊥, ¬P1 ∧ ¬P2 ∧ ... ∧ ¬Pn ∧ [])
```

Comment:
the last clause is mathematically redundant but maybe it's useful as an indicator of when the (partial) function is undefined, so keeping it has an advantage. I agree.
I agree with the design and PR is ready to merge.